File: app/src/main/java/me/devsaki/hentoid/parsers/content/AnchiraContent.java
package me.devsaki.hentoid.parsers.content;

import androidx.annotation.NonNull;

import org.jsoup.nodes.Element;

import java.util.Collections;
import java.util.List;

import javax.annotation.Nonnull;

import me.devsaki.hentoid.database.domains.AttributeMap;
import me.devsaki.hentoid.database.domains.Content;
import me.devsaki.hentoid.enums.AttributeType;
import me.devsaki.hentoid.enums.Site;
import me.devsaki.hentoid.enums.StatusContent;
import me.devsaki.hentoid.parsers.ParseHelper;
import me.devsaki.hentoid.util.StringHelper;
import pl.droidsonroids.jspoon.annotation.Selector;

public class AnchiraContent extends BaseContentParser {

    @Selector(value = "#metadata h2", defValue = NO_TITLE)
    private String title;

    @Selector(value = "#metadata span")
    private List<Element> extraData;

    @Selector(value = "#metadata a[href*='/?s=artist:']")
    private List<Element> artists;

    @Selector(value = "#metadata a[href*='/?s=parody:']")
    private List<Element> parodies;

    @Selector(value = "#metadata a[href*='/?s=tag:']")
    private List<Element> tags;

    @Selector(value = "#gallery img")
    private List<Element> imgs;


    public Content update(@NonNull final Content content, @Nonnull String url, boolean updateImages) {
        // Hitomi now uses an empty template that is populated by Javascript -> parsing is entirely done by HitomiParser
        content.setSite(Site.ANCHIRA);
        content.setRawUrl(url);
        if (null == imgs) return content.setStatus(StatusContent.IGNORED);

        content.setTitle(title);

        if (!imgs.isEmpty()) content.setCoverImageUrl(ParseHelper.getImgSrc(imgs.get(0)));

        AttributeMap attributes = new AttributeMap();
        ParseHelper.parseAttributes(attributes, AttributeType.ARTIST, artists, true, Site.ANCHIRA);
        ParseHelper.parseAttributes(attributes, AttributeType.SERIE, parodies, true, Site.ANCHIRA);
        ParseHelper.parseAttributes(attributes, AttributeType.TAG, tags, true, Site.ANCHIRA);
        content.putAttributes(attributes);

        for (Element e : extraData) {
            String txt = e.text().toLowerCase();
            if (txt.contains("page")) {
                txt = StringHelper.keepDigits(txt);
                if (StringHelper.isNumeric(txt)) {
                    content.setQtyPages(Integer.parseInt(txt));
                    break;
                }
            }
        }

        if (updateImages) {
            content.setImageFiles(Collections.emptyList());
            content.setQtyPages(0);
        }

        return content;
    }
}

File: app/src/main/java/me/devsaki/hentoid/parsers/content/AnchiraContent.kt
package me.devsaki.hentoid.parsers.content

import me.devsaki.hentoid.database.domains.AttributeMap
import me.devsaki.hentoid.database.domains.Content
import me.devsaki.hentoid.enums.AttributeType
import me.devsaki.hentoid.enums.Site
import me.devsaki.hentoid.enums.StatusContent
import me.devsaki.hentoid.parsers.cleanup
import me.devsaki.hentoid.parsers.getImgSrc
import me.devsaki.hentoid.parsers.parseAttributes
import me.devsaki.hentoid.util.isNumeric
import me.devsaki.hentoid.util.keepDigits
import org.jsoup.nodes.Element
import pl.droidsonroids.jspoon.annotation.Selector
import java.util.Locale

class AnchiraContent : BaseContentParser() {

    @Selector(value = "#metadata h2", defValue = NO_TITLE)
    private lateinit var title: String

    @Selector(value = "#metadata span")
    private var extraData: List<Element>? = null

    @Selector(value = "#metadata a[href*='/?s=artist:']")
    private var artists: List<Element>? = null

    @Selector(value = "#metadata a[href*='/?s=parody:']")
    private var parodies: List<Element>? = null

    @Selector(value = "#metadata a[href*='/?s=tag:']")
    private var tags: List<Element>? = null

    @Selector(value = "#gallery img")
    private var imgs: List<Element>? = null


    override fun update(content: Content, url: String, updateImages: Boolean): Content {
        // Hitomi now uses an empty template that is populated by Javascript -> parsing is entirely done by HitomiParser
        content.site = Site.ANCHIRA
        content.setRawUrl(url)
        if (null == imgs) return content.setStatus(StatusContent.IGNORED)
        content.title = cleanup(title)
        imgs?.let {
            if (it.isNotEmpty()) content.coverImageUrl = getImgSrc(it[0])
        }
        val attributes = AttributeMap()
        parseAttributes(attributes, AttributeType.ARTIST, artists, true, Site.ANCHIRA)
        parseAttributes(attributes, AttributeType.SERIE, parodies, true, Site.ANCHIRA)
        parseAttributes(attributes, AttributeType.TAG, tags, true, Site.ANCHIRA)
        content.putAttributes(attributes)
        var nbPages = 0
        extraData?.let {
            for (e in it) {
                var txt = e.text().lowercase(Locale.getDefault())
                if (txt.contains("page")) {
                    txt = keepDigits(txt)
                    if (isNumeric(txt)) {
                        nbPages = txt.toInt()
                        break
                    }
                }
            }
        }
        content.qtyPages = nbPages
        if (updateImages) content.setImageFiles(emptyList())

        return content
    }
}